Show `RefreshIndicator` on top when scroll's axis direction is up (matching native behaviour)

TahaTesser opened the pull request:
Fixes https://github.com/flutter/flutter/issues/26758

Created a repo for testing `RefreshIndicator` with `SwipeRefreshLayout`
https://github.com/TahaTesser/refresh_indicator_reverse_example
(Based on https://github.com/nt4f04uNd/tabs_overlay, thanks @nt4f04uNd)


### Showcase
This is commonly used in a chat app where new messages appear on top in a reversed list and when user want to load previous messages they scroll from top and indicator is shown for loading older messages (as new messages always at the bottom and these messages are usually loaded first).

Also looks like `SwipeRefreshLayout` doesn't support pull from bottom https://stackoverflow.com/questions/24867204/swiperefreshlayout-pull-from-bottom


This PR shows `RefreshIndicator` on top when scroll's axis direction is up (matching native behaviour)
### Current behaviour when `RefreshIndicator`is wrapped around reverse ListView.

| Pull from bottom  | Pull from top |
| --------------- | --------------- |
| Flutter triggers `RefreshIndicator` while Android doesn't. |  `SwipeRefreshLayout`triggered by Android. |
| <img src="https://user-images.githubusercontent.com/48603081/150970392-11fc7a70-c78c-4798-a175-6c285cb27787.gif" height="450" /> | <img src="https://user-images.githubusercontent.com/48603081/150970429-10c2e210-ad40-4a61-8a6b-85442639963c.gif" height="450" /> |

### With this PR, when `RefreshIndicator`is wrapped around reverse ListView

| Pull from bottom  | Pull from top |
| --------------- | --------------- |
|   Flutter doesn't triggers `RefreshIndicator` matching Android behaviour |  Both `RefreshIndicator` & `SwipeRefreshLayout` are triggered. |
| <img src="https://user-images.githubusercontent.com/48603081/150970666-07a4edd2-4dd3-47ec-aafe-34d2279d9b3d.gif" height="450" /> | <img src="https://user-images.githubusercontent.com/48603081/150970674-fa49c5e9-29fe-4e6e-a4eb-ad2cd92010be.gif" height="450" /> |




## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

pattaravitappsynth commented:
I want this one!!

darrenaustin commented on packages/flutter/lib/src/material/refresh_indicator.dart:
```diff
@@ -182,6 +184,11 @@ class RefreshIndicator extends StatefulWidget {
   /// else for more complicated layouts.
   final ScrollNotificationPredicate notificationPredicate;
 
+  /// Weather the refresh indicator should be positioned opposite of the scroll direction.
```
```suggestion
  /// Whether the refresh indicator should be positioned opposite of the scroll direction.
```

darrenaustin commented on packages/flutter/lib/src/material/refresh_indicator.dart:
```diff
@@ -308,7 +315,7 @@ class RefreshIndicatorState extends State<RefreshIndicator> with TickerProviderS
         indicatorAtTopNow = true;
```
Not totally following the logic here, but if the `AxisDirection.up` sets `indicatorAtTopNow` to `widget.reverse`, shouldn't this be setting it to `!widget.reverse`?

Piinks replied:
I agree, it may be easier to follow if the widget state is updated to reflect that the indicator is leading or trailing. A getter may be more readable.

TahaTesser replied:
Thanks so much for feedback! I will update this soon

darrenaustin reviewed (changes requested):
Thanks for the contribution. Looks good. Just a small typo and a question below.

It is a shame we can't have some way for this to just track the `reverse` setting on the scrollable descendent so that this just works out of the box, but this at least gives apps the ability to manually set both the scrollable and the indicator to handle the reverse case.

Piinks reviewed (commented):
> It is a shame we can't have some way for this to just track the reverse setting on the scrollable descendent so that this just works out of the box

You should be able to do this by checking the AxisDirection of the notification. That is how we do this in widgets like the Scrollbar.

It looks like the RefreshIndicator already does this though. When the ListView is reversed, the RefreshIndicator moves to the bottom, so it is always there at the leading edge of the scroll view. I think what this is trying to do is put it at the trailing edge instead.

darrenaustin commented:
> It looks like the RefreshIndicator already does this though. When the ListView is reversed, the RefreshIndicator moves to the bottom, so it is always there at the leading edge of the scroll view. I think what this is trying to do is put it at the trailing edge instead.

Yeah, it does track the scroll direction properly.  The request from #26758 was to have it appear at the top of the list if the ListView is using the `reverse`.  I was just wondering if there was a way that the indicator could pick up on this without needed a new `reverse` parameter that would manually need to be set by the app.

Which now that I think about it, the request seems a bit odd. The whole UI gesture of 'pull down to refresh' is built on the idea that new items usually appear at the top of the list, so you scroll to the top of the list and 'pull' for more. In the case of a chat app using the reverse setting, new items will appear at the bottom, so pulling down from the top is a little odd.

TahaTesser committed:
Up Axis Direction always show indicator on top matching native Android behaviour

TahaTesser commented:
@darrenaustin @Piinks 
I agree that adding parameter is not the way to go. Please checkout my updated description and new code.
Let me know if there are other changes needed to improvide this behaviour, thanks

Piinks commented on packages/flutter/test/material/refresh_indicator_test.dart:
```diff
@@ -115,7 +115,7 @@ void main() {
     );
 
-    await tester.fling(find.text('X'), const Offset(0.0, -300.0), 1000.0);
+    await tester.fling(find.text('X'), const Offset(0.0, 600.0), 1000.0);
```
I'm running this PR through internal testing now, I am concerned it may be breaking, but hopeful it won't be.

TahaTesser replied:
@darrenaustin  actually approved this (so I won't make changes to turn it into parameter based direction) 
Can you please take a look if this breaks any internal tests?

Piinks replied:
Out internal presubmit tester is having some issues, I think this is ok to land though.

Piinks reviewed (commented):
This is great - thank you for pursuing a full fidelity update here.

I have added @nt4f04uNd because they are working on a big fidelity update as well, review not necessary, just added as an FYI :)

nt4f04uNd reviewed (commented):
I feel like this is probably a case where we should provide a better behavior for users, rather than match the native behavior. It seems right to me that refresh indicator shows by default at the leading extent.

Although, https://github.com/flutter/flutter/issues/26758 seems a valid feature request and is saying they want to change this behavior, so perhaps the `reverse` (or something like `appearAtEnd`) parameter is a better way to go? Adding docs about all of this also would be good.

What do you guys think?

nt4f04uNd commented:
I also cloned you PR and tried running it with BouncingScrollPhysics, and refresh indicator didn't appear anywhere

TahaTesser commented:
@nt4f04uNd 
If fidelity is not important, I would love to add something like `appearAtEnd`

TahaTesser commented:
> I also cloned your PR and tried running it with BouncingScrollPhysics, and refresh indicator didn't appear anywhere

Are you running on the web or iOS?

TahaTesser commented:
@nt4f04uNd 
Thanks for the test, i am surprised why it is broken with BouncingScrollPhysics. 
I will take look into adding parameter instead as well

darrenaustin reviewed (approved):
This LGTM.

TahaTesser commented:
@nt4f04uNd @Piinks 
I am not in scroll physics but it looks RefreshIndicator should use `AlwaysScrollableScrollPhysics`, not `BouncingScrollPhysics` (even tho it does show indicator without this PR for some reason)

Minimal example, this never trigger overscroll.
```
      body: Center(
        child: NotificationListener<OverscrollNotification>(
          onNotification: (OverscrollNotification overscroll) {
            print('overscroll: $overscroll');
            return false;
          },
          child: ListView.builder(
            reverse: true,
            physics: const BouncingScrollPhysics(),
            itemCount: 25,
            itemBuilder: (BuildContext context, int index) {
              return  ListTile(
                title: Text('List item -  $index'),
              );
            },
          ),
        ),
      ),
```

Also, the [Troubleshooting](https://api.flutter.dev/flutter/material/RefreshIndicator-class.html) section in the documentation also suggests this and [stackoverflow](https://stackoverflow.com/questions/48035594/flutter-notificationlistener-with-scrollnotification-vs-scrollcontroller)

Piinks commented:
I think @TahaTesser's comment about BouncingScrollPhysics not being compatible by default makes sense. The docs for BouncingScrollPhysics even mention there are caveats to this type of ScrollPhysics triggering overscroll:

> [BouncingScrollPhysics](https://master-api.flutter.dev/flutter/widgets/BouncingScrollPhysics-class.html) by itself will not create an overscroll effect if the contents of the scroll view do not extend beyond the size of the viewport. To create the overscroll and bounce effect regardless of the length of your scroll view, combine with [AlwaysScrollableScrollPhysics](https://master-api.flutter.dev/flutter/widgets/AlwaysScrollableScrollPhysics-class.html).

And the docs for the refresh indicator specify you need to trigger an overscroll to trigger the refresh. We may be able to make this clearer though in the docs. 

Often on iOS you'll find that AlwaysScrollableScrollPhysics being used with BouncingScrollPhysics. :)

Piinks reviewed (commented):
I am not sure this achieves the desired result after checking it out and testing.

If I use the following sample (from the api docs, master branch: https://master-api.flutter.dev/flutter/material/RefreshIndicator-class.html), on iOS it works fine, until I set reverse true on the ListView. Then I can't make the refresh indicator appear at all. It doesn't seem right that it works for `reverse: false`, but not `reverse: true` now.
On Android, it appears to work fine both ways.

This seems like it might be broken, can you take a look @TahaTesser?

Here's the code sample:

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'RefreshIndicator Sample';

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: RefreshIndicatorExample(title: _title),
    );
  }
}

class RefreshIndicatorExample extends StatefulWidget {
  const RefreshIndicatorExample({Key? key, required this.title})
      : super(key: key);

  final String title;

  @override
  State<RefreshIndicatorExample> createState() =>
      _RefreshIndicatorExampleState();
}

class _RefreshIndicatorExampleState extends State<RefreshIndicatorExample> {
  final GlobalKey<RefreshIndicatorState> _refreshIndicatorKey =
      GlobalKey<RefreshIndicatorState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: RefreshIndicator(
        key: _refreshIndicatorKey,
        color: Colors.white,
        backgroundColor: Colors.blue,
        strokeWidth: 4.0,
        onRefresh: () async {
          // Replace this delay with the code to be executed during refresh
          // and return a Future when code finishs execution.
          return Future<void>.delayed(const Duration(seconds: 3));
        },
        // Pull from top to show refresh indicator.
        child: ListView.builder(
          reverse: true,
          itemCount: 25,
          itemBuilder: (BuildContext context, int index) {
            return ListTile(
              title: Text('Item $index'),
            );
          },
        ),
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: () {
          // Show refresh indicator programmatically on button tap.
          _refreshIndicatorKey.currentState?.show();
        },
        icon: const Icon(Icons.refresh),
        label: const Text('Show Indicator'),
      ),
    );
  }
}
```

TahaTesser committed:
fix dragoffset for ios

TahaTesser committed:
Add test for reverse BouncingScrollPhysics

skia-gold commented:
Gold has detected about 38 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/93779

TahaTesser commented:
@Piinks 
Thank you so much for taking the time to test this

Looks like I missed adjusting `_dragOffset`, this is now fixed and I also added a test for this using `BouncingScrollPhysics`.

### Preview

https://user-images.githubusercontent.com/48603081/154487546-2ff03752-0bb3-45db-9914-20e40b886d82.mov





 I've renamed some existing to match new behavior `bottom` doesn't seem right in the title as this PR would bring refresh indicator matching native behavior.

Piinks reviewed (approved):
LGTM! Thank you for the quick follow up!

AndreiMisiukevich commented:
Hello @TahaTesser

Do I understand correctly that RefreshIndicator is not supposed to work with reverse list view?
Messenger app is a pretty common use-case for upside-down pull-to-refresh behavior.
Now it seems impossible to achieve (except tricks with RotatedBox which won't be 100% working it changes axisdirection of list view).

Thanks

MattyBoy4444 commented:
> Hello @TahaTesser
> 
> Do I understand correctly that RefreshIndicator is not supposed to work with reverse list view? Messenger app is a pretty common use-case for upside-down pull-to-refresh behavior. Now it seems impossible to achieve (except tricks with RotatedBox which won't be 100% working it changes axisdirection of list view).
> 
> Thanks

I agree
